gpii/node_modules/lifecycleManager/src/LifecycleManagerSession.js

```diff
@@ -72,17 +72,55 @@ var gpii = fluid.registerNamespace("gpii");
     });
 
     /**
-     * Save prefs to the cloud when the preferences are updated. Note that change requests on session.model.preferences
+     * Filter the preferences by metadata defined in the preferences set. Only the preferences that are allowed to be
+     * autosaved will be returned.
+     * @param {Object} preferences - The updated preferences.
+     * @param {Object} metadata - The metadata that defined in the prefs set.
+     * @return {Object} - The preferences to be autosaved.
+     */
+    gpii.lifecycleManager.session.filterPrefsByAutoSave = function (preferences, metadata) {
+        var autosaveScope;
+        fluid.each(metadata, function (meta) {
```

Comment:
This is ok as a first pass but there are plenty of subtleties not handled.
i) There may be more than one "autosave" type metadata
ii) This algorithm doesn't handled "scoped preferences" which didn't exist at the time the spec was written but have now been adopted as a way of writing various kinds of preferences until we can have a fixed matchmaker.
I suggest you handle these by allowing arrays as members of "scope" - e.g. represent the path to this one https://github.com/GPII/universal/blob/master/testData/preferences/snapset_2a.json5#L25
as ["http://registry.gpii.net/applications/net.gpii.uioPlus", "http://registry.gpii.net/common/highContrast/enabled"]
just so you don't have to deal with the horror of escaping.
To make the lower algorithm clearer I suggest you construct an "isomorphic skeleton" of the preferences to be saved by using fluid.set(scope, path, true) rather than accumulating a flat list as you do. This also naturally takes care of deduplication and allows for the possibility of wildcards.

Then your check lower down just consists of testing whether fluid.get(scope, path) has a value or not.

Reply:
Cheers for the updates, @cindyli - I was turning over in bed the greater reason for this isomorphic map is that we are in practice going to have to support values of the metadata with values "true" and "false". This is because there is going to have to be a default policy in the core which automatically autosaves some preferences (that is, I predict that once Gregg understands the implications of what we have just built, he is going to ask for this) and so the user is going to be able to have to tweak this negatively as well as positively. Therefore the cleanest representation for this is a map into which we have applied the true/false values to the metadata paths using fluid.set, in order of seeing the rules.

Reply:
Shall I make more updates in this pull request to support "autosave" metadata in a structure like: 

```
"metadata": [
    {
        "type": "autosave",
        "scope": [
            "http://registry.gpii.net/common/magnification"
        ],
        "value": true
    },
    {
        "type": "autosave",
        "scope": [
            [
                "http://registry.gpii.net/applications/org.gnome.desktop.a11y.magnifier",
                "http://registry.gpii.net/common/magnification"
            ]
        ],
        "value": false
    },
    {
        "type": "autosave",
        "scope": [
            "http://registry.gpii.net/common/volume"
        ],
        "value": true
    }
    ....
]
```
If so, do you think if the `scope` array should be designed to support multiple preference paths such as` ["http://registry.gpii.net/common/volume", ["volume", "pitch"]]or only one preference path per "autosave" block?